### ISSUE TITLE
Support dynamic naming strategy for collection names [DATAMONGO-1863]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-data-mongodb#2764]
**[Dan Markhasin](https://jira.spring.io/secure/ViewProfile.jspa?name=minimizer)** opened **[DATAMONGO-1863](https://jira.spring.io/browse/DATAMONGO-1863?redirect=false)** and commented

As described in [DATAMONGO-525](https://jira.spring.io/browse/DATAMONGO-525) (https://jira.spring.io/browse/DATAMONGO-525), we are looking for a way to generate collection names dynamically to support multi-tenancy.
This can be achieved by something as simple as a prefix or by a naming strategy mechanism, simliar to FieldNamingStrategy.

For example, the most simple solution that would work (for us) is to have something like this:

```java
@Document(prefix = "${info.service_name}")
class Dog
{
private String breed;
}
```

That would generate a collection called animalService_Dog (assuming the service name is animalService).
Any other solution is welcome, but I believe this is a major feature that is currently lacking which is critical for multi-tenancy support

---

**Reference URL:** https://jira.spring.io/browse/DATAMONGO-525

1 votes, 3 watchers

**Comment by spring-issuemaster** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by alxalx22** [comment]
> Does that already do what you want?

No, we need a way to specify in `@Document` annotation the collection name as property from properties file

**Comment by harinibodipudi** [comment]
https://www.examplefiles.net/cs/1040155

This link specifies how we can use dynamic collection names with the environment variables. It worked for me

**Comment by Steve973** [comment]
I am trying to do something like this, too.  My use case entails a configuration that specifies a list of mongo urls, and/or a list of collections for the same URL.  If, in my application, I want to perform a kind of query federation over some configured number of collections in one or more mongo instances, I need a way to set up the collection name in the repositories that I configure, especially if they use the same model class, and I don't want to mandate the exact name of the collection.  Please let me know if that explanation doesn't make sense, and I will do my best to clear it up.

I can get halfway there, and I'll explain what that means.  If I disable mongo repository auto-configuration, and create the repository myself, I can construct a MongoEntityInformation object, and pass the collection name as a constructor parameter.  In a custom repo impl, I extend SimpleMongoRepository, and pass this (along with the mongo template) into the constructor.  If I use methods from the SimpleMongoRepository base class, then it uses the collection name that is set in the MongoEntityInfo instance.  But, if I have methods in the original interface (the one that you include the custom interface in the list of "implements") then it only uses the collection name derived from the model class (either the lower-camel-case version of the name, or whatever is in the @Document annotation.

I have an example [here](https://github.com/Steve973/multiple-mongo-config/) that demonstrates what I mean.  The test writes to embedded mongo, and you have to look in the log output to see the collection name.  I don't have a test that checks for the collection that is used when interface methods are invoked, but I can add that soon.

**Comment by Steve973** [comment]
> > Does that already do what you want?
> 
> No, we need a way to specify in `@Document` annotation the collection name as property from properties file

This feature has already worked for quite a while.  if you put @Document(collection = "#{@nameOfBeanHoldingSomeString}" on your model class, it will use the string value of a bean of that name as the collection name.  In a configuration class, you read in some property, and expose it as a named bean.  And that's it.  My case, mentioned above, is a bit more complicated.

**Comment by christophstrobl** [comment]
Related to: spring-data-commons/issues#2369.

**Comment by christophstrobl** [comment]
resolved via #2705 4683
